rfc/0009-policy-report.md

```diff
@@ -0,0 +1,323 @@
+More policies can be added or they can be redeployed with different settings.
+As a result of that, a Kubernetes resource that was marked as valid even a couple of hours ago,
+could be instead rejected later on because of a policy change.
+
+We want to implement a audit check that inspects all the resources that are already
+present inside of the cluster and flags all the ones that are violating at least one of
+the currently enforced policies.
+
+The results of these evaluations must be accessible by operators and end users
+in an easy way.
+
+## Examples / User Stories
+
+As a operator, I want to know if some resource in my cluster violates the latest policies.
+
+As a user, I want to know if some of resource that I own violates the cluster compliance policies.
+
+As a Kubernetes developer, I want access to the audit checks results to allow me pro-grammatically act upon it.
+
+# Detailed design
+
+Kubewarden will generate audit reports using the
+[PolicyReport](https://github.com/kubernetes-sigs/wg-policy-prototypes/tree/master/policy-report)
+format that is standardized by the Kubernetes SIG "wg-policy".
+
+The audit reports are defined as set of Kubernetes CRDs. The CRDs are pretty
+flexible, they can be used to store audit
+results from Kubernetes policy engines (like Kubewarden or Kyverno) but can
+also be used to record audit results from other kind of tools (like Kubernetes
+security scanning tools as Falco or Kubebench).
+
+Kubewarden will generate one `PolicyReport` object per each Namespace found inside of
+the cluster.
+Plus, it will maintain one single `ClusterPolicyReport` object to keep track
+of all the cluster-wide Kubernetes resources.
+
+## Namespace report
+
+A Namespace can be "affected" both by `ClusterAdmissionPolicy` and by `AdmissionPolicy` policies.
+When auditing a Namespace, only some Kubewarden policies are going to be interested
+about the events happening inside of that Namespace. In addition to that,
+the same Kubernetes resource could be evaluated by multiple policies. This will
+lead to several policy evaluations being performed.
+
+The `PolicyReport` is a namespaced CR that provides these information:
+
+* How many policy evaluations have been done during the audit run of the Namespace
+* How many policy evaluations led to a resource being accepted, rejected or
+  could not be performed because of an error.
+  Policy failures happen when an error occurs during the policy evaluation,
+  for example, because of a runtime panic
+* Detailed results about all the evaluations performed. An evaluation result contains
+  the following information: the policy being used, the resource that was
+  evaluated, the final outcome (accept/reject/failure) and the eventual message
+  returned by the policy
+
+The `PolicyReport` of the Namespace is kept updated during the lifetime of the
+cluster, meaning:
+
+* New Policies might become "interested" about the events happening inside of
```

Comment:
Policies may be written to check UPDATE operations of PolicyReport. This will trigger an endless loop of entries to the PolicyReport, and possible DOS of the policy server/cluster. Do we need to care about this? Should we disallow policies from checking wgpolicyk8s.io/v1alpha2 CRDs (that includes PolicyReports, PolicyReportResults..).

Reply:
This is really a corner case, I would not focus on that.

I think that, given `PolicyReport` are namespaced and contain security-related details, regular users should not be allowed to DELETE/UPDATE them. They should just have READ access to them. This can be done via RBAC policies. We might want to document that.

Reply:
I've extended the RFC to point this out